File: LeetCode/Normal/1-99/51.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        # 按行枚举
        col, diag, reverse_diag = [0] * n, [0] * 2 * n, [0] * 2 * n
        ret = []
        grid = [['.'] * n for _ in range(n)]
        def dfs(u:int):
            # 终点返回
            if u == n:
                ret.append(["".join(grid[i]) for i in range(n)])
                return
            for i in range(n):
                if not col[i] and not diag[u + i] and not reverse_diag[n - u + i]:
                    grid[u][i] = 'Q'
                    col[i], diag[u + i], reverse_diag[n - u + i] = 1, 1, 1
                    dfs(u + 1)
                    grid[u][i] = '.'
                    col[i], diag[u + i], reverse_diag[n - u + i] = 0, 0, 0
        dfs(0)
        return ret

File: LeetCode/Normal/1-99/51.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    vector<vector<string>> vts;
    vector<string> v;

    //左斜线 r-i为固定值 右斜线r+i为固定值 大小为2*n注意映射
    int row[9] = {0}, zx[20] = {0}, yx[20] = {0};

    void dfs(int r, int n)
    {
        if (r == n)
        {
            vts.push_back(v);
            return;
        }
        string s(n, '.');
        for (int i = 0; i < n; i++)
        {
            if (row[i] || zx[r - i + 9] || yx[r + i])
                continue;
            s[i] = 'Q';
            row[i] = zx[r - i + 9] = yx[r + i] = 1;
            v.push_back(s);
            dfs(r + 1, n);
            v.pop_back();
            row[i] = zx[r - i + 9] = yx[r + i] = 0;
            s[i] = '.';
        }
    }

    vector<vector<string>> solveNQueens(int n)
    {
        dfs(0, n);
        return vts;
    }
};